### ISSUE TITLE
Improve performance of `ShippingByWeightByTotalService.FindRecordsAsync`

### ISSUE THREAD

**Issue by Manuel-Innovapps** [nopSolutions/nopCommerce#6763]
nopCommerce version: 4.50.x +

Steps to reproduce the problem: 
* We are currently maintaining a nopcommerce instance for one of our customers with 25+ active shops.
* Each of these shops has their own shipping rates, and each shops makes heavy use of the `WeightFrom` and `WeightTo` properties.
* Currently we have 494k entries in the appropriate table
* Our customer reported, that the `ShippingMethod` page during checkout is often loading for 10+ seconds.

After investigating the issue in the azure logs, we were able to determine that the following sql is causing huge performance problems:
```sql
SELECT
	[t1].[Id],
	[t1].[LowerWeightLimit],
	[t1].[RatePerWeightUnit],
	[t1].[PercentageRateOfSubtotal],
	[t1].[AdditionalFixedCost],
	[t1].[OrderSubtotalTo],
	[t1].[OrderSubtotalFrom],
	[t1].[WeightTo],
	[t1].[WeightFrom],
	[t1].[TransitDays],
	[t1].[ShippingMethodId],
	[t1].[Zip],
	[t1].[StateProvinceId],
	[t1].[CountryId],
	[t1].[WarehouseId],
	[t1].[StoreId]
FROM
	[ShippingByWeightByTotalRecord] [t1]
ORDER BY
	[t1].[StoreId],
	[t1].[CountryId],
	[t1].[StateProvinceId],
	[t1].[Zip],
	[t1].[ShippingMethodId],
	[t1].[WeightFrom],
	[t1].[OrderSubtotalFrom]
```

Here is an image of the query performance insight for this exact query:
![image](https://github.com/nopSolutions/nopCommerce/assets/54266094/3cbddb9a-2018-4af6-8414-466f984a6965)

Let's look at the first row in the table. There we can see, that query is executed 4 times within an hour, and the total duration approximately one minute and ten seconds. If we divide that by 4, we can determine that each query ran for about 17 seconds. This is to much!

After analyzing the query above and the source code that appears to be producing this, it looks like there is currently a major performance bottleneck, as all 494k entries are fetched and then filtered in memory. It looks like after fetching these entries once, they appear to stay in memory for a few minutes. But storing all of those entries in memory doesn't sound like a good idea either.

Happy to provide you the contents of our table in private if necessary.

https://github.com/nopSolutions/nopCommerce/blob/2796b813f366cd0363c024d9b823933600565075/src/Plugins/Nop.Plugin.Shipping.FixedByWeightByTotal/Services/ShippingByWeightByTotalService.cs#L90-L92

**Comment by skoshelev** [comment]
Hi @Manuel-Innovapps. When we developed this plugin, we had no idea that it would be used with such a large amount of data, as a rule, such an extensive configuration is not required. We decided to review the current approach and make changes to the code. We will be very grateful if you send data for testing, ideally we need data from three tables: ShippingByWeightByTotalRecord, ShippingMethod and Store.
Thanks for the detailed description of the problem. You can send the data in the archive to sergey.k@nopcommerce.com or via [PM on the forum](https://www.nopcommerce.com/profile/156980), whichever is more convenient for you

**Comment by Manuel-Innovapps** [comment]
Did you had a chance to look at the data we provided? Do you need any additional informations?

**Comment by skoshelev** [comment]
Hi @Manuel-Innovapps. Yes, thank you very much, I'm working on this problem. I'm trying to find a compromise between the load on the database and on the RAM. I don't think I need any additional information, those that you gave us should be enough.

**Comment by Manuel-Innovapps** [comment]
@skoshelev Great to hear! If you have anything ready for testing, we are more than happy to cherry pick and test the changes!

**Comment by skoshelev** [comment]
Hi @Manuel-Innovapps. Even on your database in my environment, the previous approach worked well with a medium server load. However, as I said, I rewrote caching into more fractional parts, I hope that in real life it will behave well and the load on the transport level and RAM will decrease to significant values.

Closed #6763
